src/serializer/serializer.js

```diff
@@ -750,6 +750,100 @@ export class Serializer {
     return t.arrayExpression(elems);
   }
 
+  _serializeValueMap(name: string, val: ObjectValue, reasons: Array<string>): BabelNodeExpression {
+    let kind = val.getKind();
+    let elems = [];
+
+    let entries;
+    if (kind === "Map") {
+      entries = val.$MapData;
+    } else {
+      invariant(kind === "WeakMap");
+      entries = val.$WeakMapData;
+    }
+    invariant(entries !== undefined);
+    let len = entries.length;
+
+    for (let i = 0; i < len; i++) {
+      let entry = entries[i];
+      let key = entry.$Key;
+      let value = entry.$Value;
+      if (key === undefined || value === undefined) continue;
+      let mightHaveBeenDeleted = key.mightHaveBeenDeleted();
+      let delayReason = this._shouldDelayValue(key) ||
+        this._shouldDelayValue(value) || mightHaveBeenDeleted;
+        if (delayReason) {
+          // handle self recursion
+          this._delay(delayReason, [key, value, val], () => {
+            invariant(key !== undefined);
+            invariant(value !== undefined);
+            this.body.push(t.expressionStatement(
+              t.callExpression(
+                t.memberExpression(this._getValIdForReference(val), t.identifier("set")),
+                [this.serializeValue(key, reasons.concat(`Set entry on ${name}`)),
+                 this.serializeValue(value, reasons.concat(`Set entry on ${name}`))]
+              )
+            ));
+          });
+        } else {
+          let serializedKey = this.serializeValue(key, reasons);
+          let serializedValue = this.serializeValue(value, reasons.concat(`Set entry on ${name}`));
+          let elem = t.arrayExpression([serializedKey, serializedValue]);
+          elems.push(elem);
+        }
+    }
+
+    this.addProperties(name, val, reasons, val.properties);
+    let arrayValue = t.arrayExpression(elems);
+    return t.newExpression(
+      this.preludeGenerator.memoizeReference(kind), [arrayValue]);
+  }
+
+  _serializeValueSet(name: string, val: ObjectValue, reasons: Array<string>): BabelNodeExpression {
+    let kind = val.getKind();
+    let elems = [];
+
+    let entries = val.$SetData;
+    if (kind === "Set") {
+      entries = val.$SetData;
+    } else {
+      invariant(kind === "WeakSet");
+      entries = val.$WeakSetData;
+    }
+    invariant(entries !== undefined);
+    let len = entries.length;
+
+    for (let i = 0; i < len; i++) {
+      let entry = entries[i];
+      if (entry === undefined) continue;
+      let mightHaveBeenDeleted = entry.mightHaveBeenDeleted();
```

Comment:
I don't quite see how an entry ".mightHaveBeenDeleted()". Is this getting triggered by one of the test cases? If not, and you can't think of one, I suggest to rather have the invariant that !entry.mightHaveBeenDeleted.